solarforecastarbiter/datamodel.py

```diff
@@ -419,6 +419,130 @@ class Observation(BaseModel):
     __post_init__ = __set_units__
 
 
+@dataclass(frozen=True)
+class AggregateObservation(BaseModel):
+    """
+    Class for keeping track of Observations in Aggregates
+    and when they are added and removed from the Aggregate
+
+    Parameters
+    ----------
+    observation : Observation
+        The Observation object that is part of the Aggregate
+    effective_from : pandas.Timestamp
+        The effective datetime of when the Observation should be
+        included in the Aggregate
+    effective_until : pandas.Timestamp
+        The effective datetime of when the Observation should be
+        excluded from the Aggregate
+    observation_deleted_at : pandas.Timestamp
+        The datetime that the Observation was deleted from the
+        Arbiter. This indicates that the Observation should be
+        removed from the Aggregate, and without the data
+        from this Observation, the Aggregate is invalid before
+        this time.
+
+    See Also
+    --------
+    Observation
+    Aggregate
+    """
+    observation: Observation
+    effective_from: pd.Timestamp
+    effective_until: Union[pd.Timestamp, None] = None
+    observation_deleted_at: Union[pd.Timestamp, None] = None
+
+    def _special_field_processing(self, model_field, val):
+        if model_field.name in ('effective_until', 'observation_deleted_at'):
+            if val is None:
+                return val
+            else:
+                return pd.Timestamp(val)
+        else:  # pragma: no cover
+            return val
+
+
+def __check_variable__(variable, *args):
+    if not all(arg.variable == variable for arg in args if arg is not None):
+        raise ValueError('All variables must be identical.')
+
+
+def __check_aggregate_interval_compatibility__(interval, *args):
+    if any(arg.interval_length > interval for arg in args if arg is not None):
+        raise ValueError('observation.interval_length cannot be greater than '
+                         'aggregate.interval_length.')
+    if any(arg.interval_value_type not in ('interval_mean', 'instantaneous')
+           for arg in args if arg is not None):
+        raise ValueError('Only observations with interval_value_type of '
+                         'interval_mean or instantaneous are acceptable')
+
+
+@dataclass(frozen=True)
+class Aggregate(BaseModel):
+    """
+    Class for keeping track of Aggregate metadata.
+
+    Parameters
+    ----------
+    name : str
+        Name of the Aggregate, e.g. Utility X Solar PV
+    description : str
+        A description of what the aggregate is.
+    variable : str
+        Variable name, e.g. power, GHI. Each allowed variable has an
+        associated pre-defined unit. All observations that make up the
+        Aggregate must also have this variable.
+    aggregate_type : str
```

Comment:
Is this supposed to be restricted to sum and mean? Do other functions work? min, max, median are the first that come to mind.